src/writer.rs

```diff
@@ -54,23 +58,28 @@ impl<'a, W: Write> Writer<'a, W> {
         }
     }
 
+    /// Get a reference to the `Schema` associated to a `Writer`.
     pub fn schema(&self) -> &'a Schema {
         self.schema
     }
 
+    /// Append a compatible value (implementing the `ToAvro` trait) to a `Writer`, also performing
+    /// schema validation.
+    ///
+    /// Return the number of bytes written (it might be 0, see below).
+    ///
+    /// **NOTE** This function is not guaranteed to perform any actual write, since it relies on
+    /// internal buffering for performance reasons. If you want to be sure the value has been
+    /// written, then call [`flush`](struct.Writer.html#method.flush).
     pub fn append<T: ToAvro>(&mut self, value: T) -> Result<usize, Error> {
         if !self.has_header {
-            let header = header(self.schema, self.codec, self.marker.as_ref());
+            let header = header(self.schema, self.codec, self.marker.as_ref())?;
             self.append_bytes(header.as_ref())?;
```

Comment:
This line actually calls `self.writer.write(...)`, we should probably take its return value (number of bytes written) and add them to the result of this function?

I wrote the code in the first place, I just realized my mistake, feel free to open an issue instead if you think this should be part of a different commit.

Reply:
No no, I'll just amend my last one. Unit tests are there also to fix bugs.